test_agent/rust/src/testagent.rs

```diff
@@ -0,0 +1,293 @@
+                    // Handle socket errors (e.g., connection closed)
+                    dbg!("Socket error: {}", e);
+                    break;
+                }
+            };
+            if bytes_received == 0 {
+                continue;
+            }
+
+            let recv_data_str: std::borrow::Cow<'_, str> =
+                String::from_utf8_lossy(&recv_data[..bytes_received]);
+            let mut action_str = "";
+            let cleaned_json_string = sanitize_input_string(&recv_data_str).replace("BYTES:", "");
+            let json_msg: Value =
+                serde_json::from_str(&cleaned_json_string.to_string()).expect("issue in from str"); // Assuming serde_json is used for JSON serialization/deserialization
+
+            let action = json_msg["action"].clone();
+            let json_data_value = json_msg["data"].clone();
+            let test_id = json_msg["test_id"].clone();
+
+            let json_str_ref = action
+                .as_str()
+                .expect("issue in converting value to string");
+
+            dbg!(json_str_ref);
+            let status = match json_str_ref {
+                constants::SEND_COMMAND => {
+                    let wu_message: WrapperUMessage =
+                        serde_json::from_value(json_data_value).unwrap(); // convert json to UMessage
+                    let u_message = wu_message.0;
+                    action_str = constants::SEND_COMMAND;
+                    self.utransport.send(u_message).await
+                }
+
+                constants::REGISTER_LISTENER_COMMAND => {
+                    let cloned_listener = Arc::clone(&arc_self);
+
+                    let wu_uuri: WrapperUUri = serde_json::from_value(json_data_value).unwrap(); // convert json to UMessage
+                    let u_uuri = wu_uuri.0;
+                    action_str = constants::REGISTER_LISTENER_COMMAND;
+                    self.utransport
+                        .register_listener(
+                            u_uuri,
+                            Arc::clone(&cloned_listener) as Arc<dyn UListener>,
+                        )
+                        .await
+                }
+
+                constants::UNREGISTER_LISTENER_COMMAND => {
+                    let cloned_listener = Arc::clone(&arc_self);
+                    let wu_uuri: WrapperUUri = serde_json::from_value(json_data_value).unwrap(); // convert json to UMessage
+                    let u_uuri = wu_uuri.0;
+                    action_str = constants::UNREGISTER_LISTENER_COMMAND;
+                    self.utransport
+                        .unregister_listener(
+                            u_uuri,
+                            Arc::clone(&cloned_listener) as Arc<dyn UListener>, /*&cloned_listener*/
+                        )
+                        .await
+                }
+
+                _ => Ok(()),
+            };
+
+            let mut status_dict: HashMap<String, _> = HashMap::new();
+
+            match status {
```

Comment:
If I'm reading this correctly, the idea is to be able to return back to the transaction manager the status of the command we were given, right?

Perhaps something like this be a little more idiomatic:
```rust
let aligned_status = {
    match status {
        // I assume here you want to get an explicit UCode::OK back here to align with the
        // other implementations
        Ok(()) => { UStatus::ok() }
        Err(ustatus) => { ustatus }
    }
}

let transaction_status = HashMap::from([
  ("message".to_string(), aligned_transaction.get_message()),
  ("code".to_string(), aligned_transaction.get_code().value().to_string()),
]);

let json_message = JsonResponseData {
	action: action_str.to_owned(),
	data: transaction_status,
	ue: "rust".to_owned(),
	test_id: test_id.to_string(),
};
```